File: 46-Permutations/46-Permutations.java
// Last updated: 8/16/2026, 9:53:18 PM
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        // DFS - all permutations
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, res, new ArrayList<>());
        return res;
    }

    private void dfs(int[] nums, List<List<Integer>> res, List<Integer> curr) {
        if (curr.size() > nums.length) return;
        if (curr.size() == nums.length) {
            res.add(new ArrayList(curr)); // deep copy
            return;
        }

        // iterate all possible candidates
        for (int i = 0; i < nums.length; i ++) {
            // skip already added candidates 
            if (curr.contains(nums[i])) continue;
            
            curr.add(nums[i]);
            dfs(nums, res, curr);
            curr.remove(curr.size() - 1);
        }
    }
}

File: 46-Permutations/46-Permutations.py
# Last updated: 8/23/2026, 1:07:32 PM
1class Solution:
2    def permute(self, nums: List[int]) -> List[List[int]]:
3        # backtrack
4        n = len(nums)
5        res = []
6        path = []           # curr path
7        used = [False] * n  # for each idx if its used in path
8       
9        def backtrack(used: list[bool]):
10            if len(path) == n:
11                res.append(path[:])
12                return
13            
14            for i in range(n):
15                if used[i]:
16                    continue
17
18                path.append(nums[i])    # select i
19                used[i] = True
20
21                backtrack(used)
22                
23                path.pop()
24                used[i] = False         # cancel selection i
25
26
27        backtrack(used)
28        return res